src/yunohost/user.py

```diff
@@ -566,6 +608,224 @@ def user_info(username):
+                                delimiter=';', quotechar='"')
+        writer.writeheader()
+        users = user_list(CSV_FIELDNAMES)['users']
+        for username, user in users.items():
+            user['mail-alias'] = ','.join(user['mail-alias'])
+            user['mail-forward'] = ','.join(user['mail-forward'])
+            user['groups'] = ','.join(user['groups'])
+            writer.writerow(user)
+
+        body = csv_io.getvalue().rstrip()
+    if msettings.get('interface') == 'api':
+        # We return a raw bottle HTTPresponse (instead of serializable data like
+        # list/dict, ...), which is gonna be picked and used directly by moulinette
+        from bottle import HTTPResponse
+        response = HTTPResponse(body=body,
+                        headers={
+                            "Content-Disposition": "attachment; filename=users.csv",
+                            "Content-Type": "text/csv",
+                        })
+        return response
+    else:
+        return body
+
+
+@is_unit_operation()
+def user_import(operation_logger, csvfile, update=False, delete=False):
+    """
+    Import users from CSV
+
+    Keyword argument:
+        csvfile -- CSV file with columns username;firstname;lastname;password;mailbox_quota;mail;alias;forward;groups
+
+    """
+
+    import csv # CSV are needed only in this function
+    from moulinette.utils.text import random_ascii
+    from yunohost.permission import permission_sync_to_user
+    from yunohost.app import app_ssowatconf
+
+    # Pre-validate data and prepare what should be done
+    actions = {
+        'created': [],
+        'updated': [],
+        'deleted': []
+    }
+    is_well_formatted = True
+
+    def to_list(str_list):
+        return str_list.split(',') if str_list else []
+
+    existing_users = user_list()['users']
+    past_lines = []
+    reader = csv.DictReader(csvfile, delimiter=';', quotechar='"')
+    for user in reader:
+        # Validation
+        try:
+            format_errors = [key + ':' + str(user[key])
+                            for key, validator in VALIDATORS.items()
+                         if user[key] is None or not re.match(validator, user[key])]
+        except KeyError as e:
+            logger.error(m18n.n('user_import_missing_column',
+                                column=str(e)))
+            is_well_formatted = False
+            break
+
+        if 'username' in user:
+            if user['username'] in past_lines:
+                format_errors.append('username: %s (duplicated)' % user['username'])
+            past_lines.append(user['username'])
+        if format_errors:
+            logger.error(m18n.n('user_import_bad_line',
+                                line=reader.line_num,
+                                details=', '.join(format_errors)))
+            is_well_formatted = False
+            continue
+
+        # Choose what to do with this line and prepare data
+        user['groups'] = to_list(user['groups'])
```

Comment:
Don't we want to check if the groups exist here? What happens if they don't exist